File: apps/api/src/models/strava_token.py
from datetime import datetime
from uuid import UUID

from sqlalchemy import BigInteger, DateTime, ForeignKey, LargeBinary, SmallInteger, String, func
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column

from src.db.base import Base


class StravaToken(Base):
    __tablename__ = "strava_tokens"
    __table_args__ = ({"schema": "public"},)

    user_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("public.users.id", ondelete="CASCADE"),
        primary_key=True,
    )
    access_token_enc: Mapped[bytes] = mapped_column(LargeBinary, nullable=False)
    refresh_token_enc: Mapped[bytes] = mapped_column(LargeBinary, nullable=False)
    key_version: Mapped[int] = mapped_column(SmallInteger, nullable=False, default=1)
    expires_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    scope: Mapped[str] = mapped_column(String, nullable=False)
    athlete_strava_id: Mapped[int] = mapped_column(BigInteger, nullable=False, unique=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    last_used_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

File: packages/shared/src/strava-token.ts
// Mirror of public.strava_tokens from supabase/migrations/0002_strava_infra.sql
// + 0003_security_hardening.sql (adds key_version).
//
// Sensitive table:
// - Excluded from supabase_realtime publication.
// - Service-role-only writes (Next.js OAuth callback handler).
// - Self-only RLS reads (the user themselves).
//
// Encryption is Node-side AES-256-GCM (see apps/web/src/security/token-crypto.ts).
// The symmetric key never traverses SQL. `key_version` enables incremental key
// rotation: new encryptions use the highest version available; all listed
// versions are tried on decrypt; each row stamps the version it was written
// under. See AGENTS.md "Secrets".

import { z } from "zod";

// PostgreSQL BYTEA representation on the wire.
// - PostgREST returns BYTEA columns as base64-encoded strings by default.
// - supabase-js INSERT accepts Uint8Array/Buffer and serialises for transit.
// Either form is valid for the row contract; conversion happens in
// apps/web/src/security/token-crypto.ts where the bytes are decrypted.
export const EncryptedBytesSchema = z.union([
  z.string(),
  z.instanceof(Uint8Array),
]);
export type EncryptedBytes = z.infer<typeof EncryptedBytesSchema>;

// Strava athlete IDs are BIGINT in SQL but fit comfortably in JS's 53-bit
// safe integer range (~10^10 today). z.number().int() is correct; revisit
// if Strava ever issues IDs larger than 2^53.
export const StravaTokenRowSchema = z.object({
  user_id: z.string().uuid(),
  access_token_enc: EncryptedBytesSchema,
  refresh_token_enc: EncryptedBytesSchema,
  expires_at: z.string().datetime({ offset: true }),
  scope: z.string(),
  athlete_strava_id: z.number().int(),
  // SMALLINT in SQL, NOT NULL DEFAULT 1. Tracks which entry in
  // STRAVA_TOKEN_KEYS was used to encrypt this row.
  key_version: z.number().int().nonnegative(),
  created_at: z.string().datetime({ offset: true }),
  last_used_at: z.string().datetime({ offset: true }).nullable(),
});

export type StravaTokenRow = z.infer<typeof StravaTokenRowSchema>;
